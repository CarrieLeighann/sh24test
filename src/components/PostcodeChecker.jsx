import React, { useState } from "react";

const allowedLSOAPrefix = ['Southwark', 'Lambeth'];
const exceptionPostcodes = ['SH24 1AA', 'SH24 1AA'];


/* This component allows users to input their postcode 
and determine if their area is served by the company or not */
export function PostcodeChecker() {

    const [postcodeResultMessage, setPostcodeResultMessage] = useState("");

    const handleSubmit = async (event) => {
        event.preventDefault();

        const submittedPostcode = event.target.elements.postcode.value;

        if (exceptionPostcodes.includes(submittedPostcode)) {
            //no need to call api
            setPostcodeResultMessage("This postcode is within the area we serve!");
        } else {
            const postcodeCheckerURL = `http://postcodes.io/postcodes/${submittedPostcode}`;
            
            fetch(postcodeCheckerURL, {method: 'GET'}).then((response) => {
                /* the api will return a 404 response when the postcode format is not valid and throw error */
                if (response.status === 200) {
                    const parsedResponse = response.json();
                    parsedResponse.then((response) => {
                        /* this pathway assumes valid responses from the api
                         error handling could be added to respond to scenarios where this is not the case */
                        const lsoa = response.result.lsoa;
                        const LSOAPrefix = lsoa.split(" ")[0];
                        const isAllowedPostcode = allowedLSOAPrefix.includes(LSOAPrefix);
                        setPostcodeResultMessage( isAllowedPostcode ? "This postcode is within the area we serve!" : "Sorry, but we do not serve this postcode")
                    });
                } else throw new Error('Not a valid postcode');
            }).catch((error) => {
                console.log(error)
                setPostcodeResultMessage("Please enter a valid postcode");
            });
        }
    }

    return (
        <div>
            <h3>Postcode Checker</h3>
            <p> To check whether we service your postcode please enter your postcode below</p>
            <form onSubmit={handleSubmit}>
                <input type="text" name="postcode" placeholder="Enter your postcode"/>
                <button type="submit">Enter</button>
            </form>
            <p>{postcodeResultMessage}</p>
        </div>
    )
}

