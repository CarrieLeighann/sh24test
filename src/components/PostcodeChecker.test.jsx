/**
 * @jest-environment jsdom
 */

import {render, screen, waitFor} from '@testing-library/react';
import "@testing-library/jest-dom";
import userEvent from '@testing-library/user-event';
import { PostcodeChecker } from "./PostcodeChecker";

//mock fetch
global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve({ status: 200, result: { lsoa: 'Southwark 3'} }),
    status: 200,
  })
);

let container, user;

beforeAll(() => {
  user = userEvent.setup()
})

beforeEach(() => {
 container = render(<PostcodeChecker/>);
})

test('renders as expected', async () => {
  await expect(container).toMatchSnapshot();
});

test('exception postcodes are accepted', async () => {
   
  const postcodeInput = await screen.findByRole('textbox');
  const submitButton = await screen.findByText('Enter');

  await user.type(postcodeInput, 'SH24 1AA');
  user.click(submitButton);
  
  await waitFor(() => {
  expect(screen.getByText('This postcode is within the area we serve!')).toBeVisible();
  });
  await expect(container).toMatchSnapshot();
});

test('valid lsoa postcodes are accepted', async () => {
const postcodeInput = await screen.findByRole('textbox');
  const submitButton = await screen.findByText('Enter');

  await user.type(postcodeInput, 'SE1 7QD');
  user.click(submitButton);

  await waitFor(() => {
  expect(screen.getByText('This postcode is within the area we serve!')).toBeVisible();
  });

});

test('invalid lsoa postcodes are rejected', async () => {

  fetch.mockImplementationOnce(() =>  Promise.resolve({
      json: () => Promise.resolve({ status: 200, result: { lsoa: 'Bathgate 2'} }),
      status: 200,
    })
  );

  const postcodeInput = await screen.findByRole('textbox');
  const submitButton = await screen.findByText('Enter');

  await user.type(postcodeInput, 'EH48 1AS');
  user.click(submitButton);

   await waitFor(() => {
  expect(screen.getByText('Sorry, but we do not serve this postcode')).toBeVisible();
  });

});

test('invalid postcodes throw error', async () => {

  fetch.mockImplementationOnce(() => Promise.resolve({status: 404}));

  const postcodeInput = await screen.findByRole('textbox');
  const submitButton = await screen.findByText('Enter');

   await user.type(postcodeInput, 'SE1 7QD');
  user.click(submitButton);

   await waitFor(() => {
  expect(screen.getByText('Please enter a valid postcode')).toBeVisible();
  });

});