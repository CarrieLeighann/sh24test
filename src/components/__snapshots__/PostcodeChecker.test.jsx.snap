// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exception postcodes are accepted 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h3>
          Postcode Checker
        </h3>
        <p>
           To check whether we service your postcode please enter your postcode below
        </p>
        <form>
          <input
            name="postcode"
            placeholder="Enter your postcode"
            type="text"
          />
          <button
            type="submit"
          >
            Enter
          </button>
        </form>
        <p>
          This postcode is within the area we serve!
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h3>
        Postcode Checker
      </h3>
      <p>
         To check whether we service your postcode please enter your postcode below
      </p>
      <form>
        <input
          name="postcode"
          placeholder="Enter your postcode"
          type="text"
        />
        <button
          type="submit"
        >
          Enter
        </button>
      </form>
      <p>
        This postcode is within the area we serve!
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`renders as expected 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h3>
          Postcode Checker
        </h3>
        <p>
           To check whether we service your postcode please enter your postcode below
        </p>
        <form>
          <input
            name="postcode"
            placeholder="Enter your postcode"
            type="text"
          />
          <button
            type="submit"
          >
            Enter
          </button>
        </form>
        <p />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h3>
        Postcode Checker
      </h3>
      <p>
         To check whether we service your postcode please enter your postcode below
      </p>
      <form>
        <input
          name="postcode"
          placeholder="Enter your postcode"
          type="text"
        />
        <button
          type="submit"
        >
          Enter
        </button>
      </form>
      <p />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
